recipes/openai_terms.py

```diff
@@ -0,0 +1,159 @@
+    progress=("Print progress of the recipe.", "flag", "pb", bool),
+    temperature=("OpenAI temperature param", "option", "mt", float),
+    top_p=("OpenAI top_p param", "option", "tp", float),
+    best_of=("OpenAI best_of param", "option", "b", int),
+    n_batch=("Batch size to send to OpenAI", "option", "bs", int),
+    max_tokens=("Max tokens to generate", "option", "t", int),
+)
+def terms_openai_fetch(
+    query: str,
+    output_path: Path,
+    seeds: List[str] = [],
+    n: int = 100,
+    model: str = "text-davinci-003",
+    prompt_path: Path = DEFAULT_PROMPT_PATH,
+    verbose: bool = False,
+    resume: bool = False,
+    progress: bool = False,
+    temperature=1.0,
+    top_p=1.0,
+    best_of=10,
+    n_batch=10,
+    max_tokens=100,
+):
+    """Get bulk term suggestions from an OpenAI API, using zero-shot learning.
+    The results can then be corrected using the `terms.openai.correct` recipe.
+
+    This approach lets you get the openai queries out of the way upfront, which can help
+    if you want to use multiple annotators of if you want to make sure you don't have to
+    wait on the OpenAI queries. The downside is that you can't flag examples to be integrated
+    into the prompt during the annotation.
+    """
+    tic = time.time()
+    template = _load_template(prompt_path)
+    # The `best_of` param cannot be less than the amount we batch.
+    if best_of < n_batch:
+        best_of = n_batch
+    
+    # Start collection of terms. If we resume we also fill seed terms with file contents.
+    terms = []
+    if resume:
+        if output_path.exists():
+            examples = srsly.read_jsonl(output_path)
+            seeds.extend([e["text"] for e in examples])
+    
+    headers = {
+        "Authorization": f"Bearer {os.getenv('OPENAI_KEY')}",
+        "OpenAI-Organization": os.getenv("OPENAI_ORG"),
+        "Content-Type": "application/json",
+    }
+
+    # This recipe may overshoot the target, but we keep going until we have at least `n`
+    while len(terms) < n:
+        prompt = template.render(n=n, examples=seeds + terms, description=query)
+        if verbose:
+            rich.print(Panel(prompt, title="Prompt to OpenAI"))
+        try:
+            resp = httpx.post(
+                "https://api.openai.com/v1/completions",
+                headers=headers,
+                json={
+                    "model": model,
+                    "prompt": [prompt],
+                    "temperature": temperature,
+                    "max_tokens": max_tokens,
+                    "top_p": top_p,
+                    "n": min(n_batch, best_of),
+                    "best_of": best_of
+                },
+                timeout=30,
+            )
+            choices = resp.json()['choices']
+        except KeyError:
+            msg.fail("Did not receive completions from OpenAI. Instead we received:")
+            rich.print(resp.json())
+            exit(code=1)
```

Comment:
This is catching `choices` not being available in the response json, right?

An `if` statement over the response code would be better here.

Reply:
Fair call. Made the change!